flask_project/campaign_manager/views.py

```diff
@@ -736,80 +979,113 @@ def create_campaign():
+def edit_campaign(mode, uuid):
+    import uuid as ud
     from flask import url_for, redirect
     from campaign_manager.forms.campaign import CampaignForm
-    from campaign_manager.models.campaign import Campaign
     """Get campaign details.
     """
+    context = {}
     try:
-        campaign = Campaign.get(uuid)
-        context = campaign.to_dict()
+        campaign = session.query(Campaign).filter(Campaign.uuid ==
+                                                  uuid).first()
+        context['campaign'] = campaign
         # Get managers
-        managers = get_allowed_managers()
-
+        managers = campaign.users
+        managers = [x.osm_user_id for x in managers]
+        campaign_types = {}
+
+        # Preprocess campaign data
+        for _ in campaign.featureTypes:
+            campaign_types[_.name] = {}
+            campaign_types[_.name]["type"] = _.name
+            campaign_types[_.name]["feature"] = _.feature
+            campaign_types[_.name]["tags"] = {}
+            for x in _.attributes:
+                campaign_types[_.name]['tags'][x.attribute_name] = []
         if request.method == 'GET':
             form = CampaignForm()
             form.name.data = campaign.name
-            form.campaign_managers.data = campaign.campaign_managers
+            form.campaign_managers.data = managers
             form.remote_projects.data = campaign.remote_projects
-            form.types.data = campaign.types
+            form.types.data = campaign_types
             form.description.data = campaign.description
-            form.geometry.data = json.dumps(campaign.geometry)
+            form.geometry.data = json.dumps(get_campaign_geometry(campaign))
             form.map_type.data = campaign.map_type
-            form.selected_functions.data = json.dumps(
-                campaign.selected_functions)
-            form.start_date.data = datetime.datetime.strptime(
-                campaign.start_date, '%Y-%m-%d')
+            form.selected_functions.data = json.dumps(get_campaign_function(
+                                                        campaign.functions))
+            form.start_date.data = campaign.start_date
             if campaign.end_date:
-                form.end_date.data = datetime.datetime.strptime(
-                    campaign.end_date, '%Y-%m-%d')
+                form.end_date.data = campaign.end_date
         else:
             form = CampaignForm(request.form)
             if form.validate_on_submit():
                 data = form.data
                 data.pop('types_options')
                 data.pop('csrf_token')
                 data.pop('submit')
-                campaign.update_data(data, form.uploader.data)
-
+                # Edit Campaign
+                if mode == "edit":
+                    fillCampaignData(campaign, data)
+                    uuid = campaign.uuid
+                # Copy Existing Campaign
+                elif mode == "copy":
```

Comment:
You should separate 'edit' and 'copy' feature.
One function should do only thing. Creation of function are cheap, use them, re-use them.
It will make the code easier to understand and to test.

Reply:
IMO your views function (or controllers) are doing too much, they should call tiny methods on models that do the actual job. I can a lot as well of code repetition.

A view function is called from a request, it delegates the job to someone else (other functions) and depending on the result, you render an approriate Response. 
In Ruby on Rails, I used to follow 'skinny controller, fat models'.

I agree that your code works and the feature you developed do what we want but IMO, coding is much more than that. I'll be happy to answer all your questions.